server/market/orderrouter.go

```diff
@@ -0,0 +1,556 @@
+			fmt.Sprintf("invalid side value %d", trade.Side))
+	}
+	quote, found := r.assets[prefix.Quote]
+	if !found {
+		panic("missing quote asset for known market should be impossible")
+	}
+	base, found := r.assets[prefix.Base]
+	if !found {
+		panic("missing base asset for known market should be impossible")
+	}
+	coins := &assetSet{
+		quote:     quote,
+		base:      base,
+		funding:   quote,
+		receiving: base,
+	}
+	if sell {
+		coins.funding, coins.receiving = base, quote
+	}
+	return tunnel, coins, sell, nil
+}
+
+// checkTimes validates the timestamps in an order prefix.
+func checkTimes(prefix *msgjson.Prefix) *msgjson.Error {
+	offset := time.Now().Unix() - int64(prefix.ClientTime)
+	if offset < 0 {
+		offset *= -1
+	}
+	if offset >= maxClockOffset {
+		return msgjson.NewError(msgjson.ClockRangeError, fmt.Sprintf(
+			"clock offset of %d seconds is larger than maximum allowed, %d seconds",
+			offset, maxClockOffset,
+		))
+	}
+	// Server time should be unset.
+	if prefix.ServerTime != 0 {
+		return msgjson.NewError(msgjson.OrderParameterError, "non-zero server time not allowed")
+	}
+	return nil
+}
+
+// checkPrefixTrade validates the information in the prefix and trade portions
+// of an order.
+func (r *OrderRouter) checkPrefixTrade(user account.AccountID, tunnel MarketTunnel, coins *assetSet, prefix *msgjson.Prefix,
+	trade *msgjson.Trade, checkLot bool) (uint64, uint32, []order.Outpoint, *msgjson.Error) {
+	// Check that the client's timestamp is still valid.
+	rpcErr := checkTimes(prefix)
+	if rpcErr != nil {
+		return 0, 0, nil, rpcErr
+	}
+
+	errSet := func(code int, message string) (uint64, uint32, []order.Outpoint, *msgjson.Error) {
+		return 0, 0, nil, msgjson.NewError(code, message)
+	}
+	// Quantity cannot be zero, and must be an integral multiple of the lot size.
+	if trade.Quantity == 0 {
+		return errSet(msgjson.OrderParameterError, "zero quantity not allowed")
+	}
+	if checkLot && trade.Quantity%coins.base.LotSize != 0 {
+		return errSet(msgjson.OrderParameterError, "order quantity not a multiple of lot size")
+	}
+	// Validate UTXOs
+	// Check that all required arrays are of equal length.
+	if len(trade.UTXOs) == 0 {
+		return errSet(msgjson.FundingError, "order must specify utxos")
+	}
+	var valSum uint64
+	var spendSize uint32
+	var utxos []order.Outpoint
+	for i, utxo := range trade.UTXOs {
+		sigCount := len(utxo.Sigs)
+		if sigCount == 0 {
+			return errSet(msgjson.SignatureError, fmt.Sprintf("no signature for utxo %d", i))
+		}
+		if len(utxo.PubKeys) != sigCount {
+			return errSet(msgjson.OrderParameterError, fmt.Sprintf(
+				"pubkey count %d not equal to signature count %d for utxo %d",
+				len(utxo.PubKeys), sigCount, i,
+			))
+		}
+		txid := utxo.TxID.String()
+		// Check that the outpoint isn't locked.
+		locked := tunnel.OutpointLocked(txid, utxo.Vout)
+		if locked {
+			return errSet(msgjson.FundingError,
+				fmt.Sprintf("utxo %s:%d is locked", utxo.TxID.String(), utxo.Vout))
+		}
```

Comment:
Could be a funding UTXO in an epoch order, and order book order, or any order with an active match. This check could also be handled by storage, since it's all tracked there too.